rero_ils/dojson/utils.py

```diff
@@ -366,3 +373,371 @@ def get_script_from_lang(asian=False):
+                self.build_part(part_number=self.part_number_waiting_name)
+            )
+        return self.part_list
+
+
+def extract_subtitle_and_parallel_titles_from_field_245_b(
+        parallel_title_data, field_245_a_end_with_equal):
+    """Extracts subtitle and parallel titles from field 245 $b.
+
+    This function retrieves the subtitle and the parallel title list
+    from field 245 $b. It also constructs a set containing the parallel title
+    and these same title without the initial article.
+    This set can be used to process the fields 246 to discard those fields
+    that match one of the elements in the set.
+    It should be noted that the deletion of the article is achieved simply
+    by deleting the first word (max 4 chars) of the parallel title.
+    It is not a problem if this word is not a real article because the chain
+    produced will not correspond to one of the 246 fields to be discarded.
+
+    :param: parallel_title_data as JSON list
+    :param: field_245_a_end_with_equal as boolean flag
+    :returns: a tuple
+     - the main subtitle
+     - a list of parallel titles
+     - a set of pararalel title strings
+    """
+
+    def remove_leading_article(string, max_article_len=4):
+        """Remove the leading article.
+
+        - Any leading word up to 'max_article_len' chars is considered
+          as an article to remove.
+        - The separateur is an APOSTROPHE or a SPACE.
+        :param string: the string to process
+        :param max_article_len: the length of article to remove (default: 4)
+        :returns: the given string without the first word
+                  with the given 'max_article_len' chars
+        :returns: an empty string otherwise
+        """
+        last_rest = ''
+        for sep in ("'", ' '):
+            first, sep, rest = string.partition(sep)
+            len_rest = len(rest)
+            if len(first) <= max_article_len \
+               and len_rest > 0 and len_rest > len(last_rest):
+                last_rest = rest
+        return last_rest
+
+    data_std = ''
+    data_lang = ''
+    lang = ''
+    main_subtitle = []
+    parallel_titles = []
+    pararalel_title_data_list = []
+    pararalel_title_string_set = set()
+    for parallel_title_value in parallel_title_data:
+        value = parallel_title_value.get('value', '')
+        lang = parallel_title_value.get('language', '')
+        if lang:
+            data_lang = value
+        else:
+            data_std = value
+
+    data_std_items = data_std.split('=')
+    data_lang_items = []
+    if data_lang:
+        data_lang_items = data_lang.split('=')
+    index = 0
+    out_data_dict = {}
+
+    for data_std in data_std_items:
+        if index == 0 and not field_245_a_end_with_equal:
+            main_subtitle.append(
+                {'value': data_std.rstrip()})
+            if lang:
+                main_subtitle.append({
+                    'value': data_lang_items[index].rstrip(),
```

Comment:
Maybe test index exist in data_lang_items.

Reply:
Ok, done